dpe/src/commands/derive_child.rs

```diff
@@ -15,36 +15,120 @@ pub struct DeriveChildCmd {
     }
 
-    const fn _is_retain_parent(&self) -> bool {
-        self.flags & Self::_RETAIN_PARENT != 0
+    const fn is_retain_parent(&self) -> bool {
+        self.flags & Self::RETAIN_PARENT != 0
     }
 
-    const fn _is_target_is_default(&self) -> bool {
-        self.flags & Self::_TARGET_IS_DEFAULT != 0
+    const fn is_make_default(&self) -> bool {
+        self.flags & Self::MAKE_DEFAULT != 0
+    }
+
+    const fn is_change_locality(&self) -> bool {
+        self.flags & Self::CHANGE_LOCALITY != 0
+    }
+
+    /// Check if it will be able to create a new default.
+    ///
+    /// # Arguments
+    ///
+    /// * `parent_idx` - Index of the soon-to-be parent.
+    /// * `default_context_idx` - Index of the current default context, if there is one.
+    fn safe_to_make_default(&self, parent_idx: usize, default_context_idx: Option<usize>) -> bool {
+        if let Some(default_idx) = default_context_idx {
+            if default_idx != parent_idx || self.is_retain_parent() {
+                return false;
+            }
+        }
+        true
     }
 }
 
 impl<C: Crypto> CommandExecution<C> for DeriveChildCmd {
-    fn execute(&self, _dpe: &mut DpeInstance<C>, _locality: u32) -> Result<Response, DpeErrorCode> {
-        Err(DpeErrorCode::InvalidCommand)
+    fn execute(&self, dpe: &mut DpeInstance<C>, locality: u32) -> Result<Response, DpeErrorCode> {
+        // Make sure the operation is supported.
+        if !dpe.support.internal_info && self.is_internal_input_info()
+            || !dpe.support.internal_dice && self.is_internal_input_dice()
+        {
+            return Err(DpeErrorCode::InvalidArgument);
+        }
+
+        let parent_idx = dpe
+            .get_active_context_pos(&self.handle, locality)
+            .ok_or(DpeErrorCode::InvalidHandle)?;
+        let child_idx = dpe
+            .get_next_inactive_context_pos()
+            .ok_or(DpeErrorCode::MaxTcis)?;
+
+        let target_locality = if !self.is_change_locality() {
+            locality
+        } else {
+            // Make sure the target locality is valid.
+            if !dpe.localities.iter().any(|&l| l == self.target_locality) {
+                return Err(DpeErrorCode::InvalidLocality);
+            }
+            self.target_locality
+        };
+
+        // Make sure it can be the default if it is supposed to be.
+        if self.is_make_default() {
+            let default_context_idx =
+                dpe.get_active_context_pos(&ContextHandle::default(), target_locality);
+
+            if !self.safe_to_make_default(parent_idx, default_context_idx) {
+                return Err(DpeErrorCode::InvalidArgument);
+            }
+        }
+
+        dpe.add_tci_measurement(child_idx, &TciMeasurement(self.data))?;
+
+        let child_handle = if self.is_make_default() {
+            ContextHandle::default()
+        } else {
+            dpe.generate_new_handle()?
+        };
+
+        let reported_parent_handle = if !self.is_retain_parent() {
```

Comment:
Optional:

```
let parent = &mut dpe.contexts[parent_idx];

if !self.is_retain_parent() {
  parent.state = ContextState::Retired;
  parent.handle = ContextHandle([0xff; ContextHandle::SIZE]);
} else if !parent.handle.is_default() {
  parent.handle = dpe.generate_new_handle()?;
}
```

Reply:
That is a lot cleaner. I had to keep the indexing for the `else` otherwise it would have 2 mutable references to something in the DPE instance.